src/main/java/htsjdk/io/AsyncWriterPool.java

```diff
@@ -0,0 +1,270 @@
+                    item = this.queue.poll();
+                }
+                return null;
+            });
+        }
+
+        /**
+         * Non-blocking close the the enclosed writer. This wraps a call to {@link PooledWriter#blockingClose()} and
+         * sends it to execute on the executor. {@code .isClosed} is immediately set to {@code true} on the current
+         * thread. The new thread will block until any ongoing writes are complete, drain any remaining enqueued items,
+         * check for exceptions, and then call close on the underlying writer.
+         *
+         * @return a future that can be waited on
+         * @throws IllegalStateException if isClosed is already true
+         */
+        private CompletableFuture<Void> nonBlockingClose() {
+            if (this.isClosed) throw new IllegalStateException("Writer is already closed");
+            this.isClosed = true;
+            return CompletableFuture.supplyAsync(() -> {
+                try {
+                    this.blockingClose();
+                    return null;
+                } catch (Exception e) {
+                    throw new RuntimeException("Caught exception while closing PooledWriter.", e);
+                }
+            }, AsyncWriterPool.this.executor);
+        }
+
+        /**
+         * Blocking close method that will first blocking and waiting for any ongoing writing to complete, draining any
+         * remaining elements from the queue, checking the result for exceptions / blocking till it is complete, and
+         * finally closing the writer itself.
+         * <p>
+         * This method should not be called directly, either {@link PooledWriter#close} or
+         * {@link PooledWriter#nonBlockingClose} should be used instead.
+         *
+         * @throws IOException if the enclosed writer's close raises an exception
+         */
+        private void blockingClose() throws IOException {
+            // Wait for any ongoing writes to finish and check for errors
+            this.blockingCheckAndRethrow();
+            // drain any remaining items in the queue
+            this.drain();
+            // Wait for last write to finish
+            this.blockingCheckAndRethrow();
+            this.writer.close();
+        }
+
+
+        /**
+         * Close the enclosed writer, first blocking and waiting for any ongoing writing to complete, draining any
+         * remaining elements from the queue, checking the result for exceptions / blocking till it is complete, and
+         * finally closing the writer itself.
+         *
+         * @throws IOException           if the enclosed writer's close raises an exception
+         * @throws IllegalStateException if isClosed is already true
+         */
+        @Override
+        public void close() throws IOException {
+            if (this.isClosed) throw new IllegalStateException("Writer is already closed");
+
+            this.isClosed = true;
```

Comment:
```suggestion
```

I would just have this call `blockingClose()` and move any checks in there as I think that makes blockingClose() and nonBlockingClose() more symmetrical.